[FIX] crm_livechat: properly handle multi-company when creating lead

seb-odoo opened the pull request:
If the current user and the selected sales team have different company, an error is raised when attempting to create the lead.

- If both partner and sales team have company, they must match or the team can't be assigned (it is more important to create the lead, which must be with the company of the partner, implying with no team which is fine as auto-assignation will take care of it).
- If one or the other has company, the lead must have this company.
- When both have no company, the lead can have no company.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/196301.png)](https://mergebot.odoo.com/odoo/odoo/pull/196301)

phenix-factory commented on addons/crm_livechat/models/chatbot_script_step.py:
```diff
@@ -26,12 +26,16 @@ def _compute_is_forward_operator(self):
         return {
-            "company_id": self.crm_team_id.company_id.id,
+            "company_id": team.company_id.id or partner.company_id.id,
             'description': description + discuss_channel._get_channel_history(),
             "name": self.env._("%s's New Lead", self.chatbot_script_id.title),
             'source_id': self.chatbot_script_id.source_id.id,
-            'team_id': self.crm_team_id.id,
+            "team_id": team.id,
```
You should change all `self.crm_team_id` to `team` or none of them.

seb-odoo replied:
I think we should actually not assign a `type` if no team is selected.

tde-banana-odoo commented on addons/crm_livechat/models/chatbot_script_step.py:
```diff
@@ -26,12 +26,16 @@ def _compute_is_forward_operator(self):
+            team = self.env["crm.team"]
         return {
-            "company_id": self.crm_team_id.company_id.id,
+            "company_id": team.company_id.id or partner.company_id.id,
```
https://github.com/odoo/odoo/pull/190364#discussion_r1903912576

Did not take long. Set to False, we generally don't want a forced company in automatically generated leads.

tde-banana-odoo replied:
Alternatively let the compute do its job, and if there is an issue maybe the compute should be updated.

seb-odoo replied:
I guess the compute should be updated then because if I set it to `False` the tests (or actual flow) won't pass.

If the partner has a company the only way to assign the lead without a crash is to have the same company on the lead.

If the team has a company the only way to assign the lead without a crash is to have the same company on the lead.

seb-odoo replied:
- “Testing Bot's New Lead” belongs to company “False” and “Sales Team” (team_id: 'Test Sale Team 2') belongs to another company.

seb-odoo replied:
Ok I got it, the compute was not called because another line (not in previous diff) was also assigning the company.
Now I removed both of them and only let the compute work, which seems to be fine after all.

tde-banana-odoo commented on addons/crm_livechat/models/chatbot_script_step.py:
```diff
@@ -26,12 +26,16 @@ def _compute_is_forward_operator(self):
     def _chatbot_crm_prepare_lead_values(self, discuss_channel, description):
+        partner = self.env.user.partner_id
+        team = self.crm_team_id
+        if partner.company_id and team.company_id and partner.company_id != team.company_id:
+            team = self.env["crm.team"]
```
That's not the right move. Team is important, company is not.

seb-odoo replied:
It's impossible to fix it differently. The only possible way to assign the partner (with a company) on the lead is to have the company match.

seb-odoo replied:
After checking again we don't have many other ways to solve it.

Main issue: we can't assign a partner and a team which have different companies. I take that for granted as I don't think we will change multi-company rules here.

We can't remove the company from the partner.
We can't remove the company from the team.

We could not assign the partner and keep the team, but it feels very counter-intuitive to be able to link the lead to an existing partner and not do it, just to preserve the team.

So the best solution for me is to assign the partner and ignore the team in this very specific case.

tde-banana-odoo commented on addons/crm_livechat/models/chatbot_script_step.py:
```diff
@@ -90,6 +94,11 @@ def _process_step_create_lead_and_forward(self, discuss_channel):
                 lambda team: team.assignment_max
                 and lead.filtered_domain(literal_eval(team.assignment_domain or "[]"))
             )
+        if self.env.user.partner_id.company_id:
```
it is probably a bad idea to create a lead, then change assignment right after that. You are going to have company / team / user mismatch, especially you globally try to fix all issues we had in base crm module.

My suggestion would be to have a common "create values" method, then create the lead in both methods but once. Don't create then change it, just create it as wanted directly.

seb-odoo replied:
How could we apply the filter on `assignment_domain` on a non-created lead?
I don't think it's possible as it possibly depends on computed field that are correctly set only after the lead is created.

seb-odoo replied:
Note: we likely never change assignment here, even though the code makes it look like it is possible.

Either the script has a forced team, and we'll use it almost no matter what (only exception: multi-company rule, in which case the team won't be previously assigned).

Or the script doesn't have a team, and it will be assigned once, at the end of this method, if an available user is found.

seb-odoo replied:
Basically this new line could be moved into the `if not teams:` if that makes it more clear, as the same check is already done in `_process_step_create_lead`.

tde-banana-odoo commented:
Globally, everything should be removed / simplified to use most of standard model compute methods. Lead model should handle it cleanly, if not, please give me real use cases. We tailored various compute fields to avoid issues like that, so you should not have to write so much layers of self rewriting code.

Cheers,

tde-banana-odoo commented:
@seb-odoo Too much comments :sweat_smile:  concerning having a partner with a company and a team with another company: indeed you are in troubles, that is notably why we try to remove as much company-based partners flows as possible (aka: don't create partners with company by default, ...). But here you have no choice as it is basically impossible, so yes I suppose the partner wins over the team.

seb-odoo commented:
@tde-banana-odoo that's a good summary.

I removed the unnecessary `company_id` in code and kept the check on team <> partner.
The branch is green included new tests I added for coverage.

Can you approve the PR?

tde-banana-odoo commented on addons/crm_livechat/models/chatbot_script_step.py:
```diff
@@ -90,6 +92,11 @@ def _process_step_create_lead_and_forward(self, discuss_channel):
                 lambda team: team.assignment_max
                 and lead.filtered_domain(literal_eval(team.assignment_domain or "[]"))
             )
+        if self.env.user.partner_id.company_id:
```
Can't you use ``_get_default_team_id`` defined in 'sales_team' module (here and/or above) ? a bit same reasoning, we try to avoid as much "custom search" as possible when searching for team.

seb-odoo replied:
It does not look like the right method for this job.

The first part of this method checks for the current user which is not what we want in this case (the current user is the visitor here, not the sales person).

And the user param of the method is the target sales person but at this step in the flow we have not assigned a sales person yet, we're only trying to find which teams are "possible", and then we select an available person from those teams based on availability for live chat.

On top of that, we want to follow assignment_domain/assignment_optout which the "default" method ignores.

tde-banana-odoo replied:
Actually it was used for company filtering, not user, read a bit fast :) .

tde-banana-odoo reviewed (commented):
Zboing zboing !

tde-banana-odoo commented on addons/crm_livechat/tests/test_chatbot_lead.py:
```diff
@@ -95,6 +95,80 @@ def test_chatbot_create_lead_portal_user(self):
+        partner = self.user_portal.partner_id
+        self.authenticate(self.user_portal.login, self.user_portal.login)
+
+        def get_lead():
```
This seems quite overkill. Can't you use RecordCapturer and simply use created lead ? Don't see why we have to reset leads at each step.

seb-odoo replied:
I guess the confusion comes from the name but the goal is not the "reset" the lead, the goal is to re-do the whole flow to test the different branching.

tde-banana-odoo commented on addons/crm_livechat/tests/test_chatbot_lead.py:
```diff
@@ -95,6 +95,80 @@ def test_chatbot_create_lead_portal_user(self):
+        partner = self.user_portal.partner_id
+        self.authenticate(self.user_portal.login, self.user_portal.login)
+
+        def get_lead():
```
"get lead" means about nothing, could you use meaningful method names ?

seb-odoo replied:
Done

seb-odoo commented:
@robodoo r+
